### PR TITLE
Observations

### PR DESCRIPTION
**Pull request by jtanx**

### CODE REVIEW

File: code/src/modules/object_tracker.cpp [hunk 1]
```diff
@@ -143,45 +158,144 @@ void ObjectTracker::Run(FlightController *fc, void *opts) {
...
+            //build observations for each
+            std::vector<Observation> visibles; //things we can currently see
+            visibles.reserve(locations.size()); //save multiple reallocations
+            for(uint i=0; i<locations.size(); i++){
+                visibles.push_back(ObservationFromImageCoords(loop_start, &gps_position, &gimbal, &imu_data, &detected_object));
+            }
+
+            if(print_observation_map){
+                //print the observations and objects
+                std::string filename = "tracker"+(static_cast<std::ostringstream*>( &(std::ostringstream() << observation_map_count++) )->str())+".png";
```
Comment (jtanx) [inline code/src/modules/object_tracker.cpp]: Can be replaced with:
std::string filename = "tracker" + std::to_string(observation_map_count++) + ".png"

**Comment by jtanx** [comment]
~~~~~
rogram received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffdcfca700 (LWP 3322)]
0x000000000052e2ea in picopter::Observations::lastObservation (this=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/observations.cpp:67
67	    return last_sample;
(gdb) bt
#0  0x000000000052e2ea in picopter::Observations::lastObservation (this=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/observations.cpp:67
#1  0x000000000052753b in picopter::ObjectTracker::Run (this=0x2129018, 
    fc=0xf9e3c0, opts=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/object_tracker.cpp:233
#2  0x0000000000509651 in picopter::FlightController::__lambda4::operator() (
    __closure=0xf9eca0)
    at /home/jeremy/fyp/picopterx/code/src/base/flightcontroller.cpp:235
#3  0x000000000050bbd2 in std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>::_M_invoke<>(std::_Index_tuple<>) (this=0xf9eca0)
    at /usr/include/c++/4.8/functional:1732
#4  0x000000000050b9d1 in std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>::operator()(void) (this=0xf9eca0) at /usr/include/c++/4.8/functional:1720
#5  0x000000000050b710 in std::_Function_handler<void(), std::reference_wrapper<std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()> > >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/4.8/functional:2101
#6  0x0000000000510778 in std::function<void ()>::operator()() const (
    this=0x21a2008) at /usr/include/c++/4.8/functional:2471
#7  0x00000000005105c2 in std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, void>---Type <return> to continue, or q <return> to quit---
::operator() (this=0x21a2000) at /usr/include/c++/4.8/future:1235
#8  0x00000000005103f1 in std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, void> >::_M_invoke(std::_Any_data const&) (
    __functor=...) at /usr/include/c++/4.8/functional:2057
#9  0x000000000050da8f in std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>::operator()() const (
    this=0x7fffdcfc9ae0) at /usr/include/c++/4.8/functional:2471
#10 0x000000000050d0e0 in std::__future_base::_State_base::_M_do_set(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&) (this=0xf9ec18, __f=..., 
    __set=@0x7fffdcfc9a7f: false) at /usr/include/c++/4.8/future:471
#11 0x000000000050fc95 in std::_Mem_fn<void (std::__future_base::_State_base::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&)>::operator()<std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >, std::reference_wrapper<bool>, void>(std::__future_base::_State_base*, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >&&, std::reference_wrapper<bool>&&) const (this=0x7fffdcfc9a08, 
    __object=0xf9ec18) at /usr/include/c++/4.8/functional:601
#12 0x000000000050f9ed in std::_Bind_simple<std::_Mem_fn<void (std::__future_bas---Type <return> to continue, or q <return> to quit---
e::_State_base::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&)> (std::__future_base::_State_base*, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >, std::reference_wrapper<bool>)>::_M_invoke<0ul, 1ul, 2ul>(std::_Index_tuple<0ul, 1ul, 2ul>) (this=0x7fffdcfc99f0) at /usr/include/c++/4.8/functional:1732
#13 0x000000000050f4f5 in std::_Bind_simple<std::_Mem_fn<void (std::__future_base::_State_base::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&)> (std::__future_base::_State_base*, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >, std::reference_wrapper<bool>)>::operator()() (this=0x7fffdcfc99f0)
    at /usr/include/c++/4.8/functional:1720
#14 0x000000000050ee51 in std::__once_call_impl<std::_Bind_simple<std::_Mem_fn<void (std::__future_base::_State_base::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&)> (std::__future_base::_State_base*, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >, std::reference_wrapper<bool>)> >() ()
    at /usr/include/c++/4.8/mutex:754
#15 0x00007ffff6134a90 in pthread_once ()
    at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_once.S:103
#16 0x0000000000508d68 in __gthread_once (__once=0xf9ec84, 
---Type <return> to continue, or q <return> to quit---
    __func=0x4beff0 <__once_proxy@plt>)
    at /usr/include/x86_64-linux-gnu/c++/4.8/bits/gthr-default.h:699
#17 0x000000000050d9d1 in std::call_once<void (std::__future_base::_State_base::*)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&), std::__future_base::_State_base* const, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >, std::reference_wrapper<bool> >(std::once_flag&, void (std::__future_base::_State_base::*&&)(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&), std::__future_base::_State_base* const&&, std::reference_wrapper<std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()> >&&, std::reference_wrapper<bool>&&) (__once=..., 
    __f=<unknown type in /home/jeremy/fyp/picopterx/code/bin/picopter, CU 0xdb994, DIE 0x10f62a>) at /usr/include/c++/4.8/mutex:786
#18 0x000000000050d04c in std::__future_base::_State_base::_M_set_result(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>, bool) (this=0xf9ec18, __res=..., 
    __ignore_failure=false) at /usr/include/c++/4.8/future:358
#19 0x000000000050afc1 in std::__future_base::_Async_state_impl<std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::{lambda()#1} ()>, void>::_Async_state_impl(picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter---Type <return> to continue, or q <return> to quit---
::FlightTask>, void*)::{lambda()#1} (&&)())::{lambda()#1}::operator()() const
    () at /usr/include/c++/4.8/future:1498
#20 0x000000000050cab0 in std::_Bind_simple<std::__future_base::_Async_state_impl<_BoundFn, <template-parameter-1-2> >::_Async_state_impl(_BoundFn&&) [with _BoundFn = std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>; _Res = void]::__lambda3()>::_M_invoke<>(std::_Index_tuple<>) (this=0xf82d70)
    at /usr/include/c++/4.8/functional:1732
#21 0x000000000050c973 in std::_Bind_simple<std::__future_base::_Async_state_impl<_BoundFn, <template-parameter-1-2> >::_Async_state_impl(_BoundFn&&) [with _BoundFn = std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>; _Res = void]::__lambda3()>::operator()(void) (this=0xf82d70)
    at /usr/include/c++/4.8/functional:1720
#22 0x000000000050c572 in std::thread::_Impl<std::_Bind_simple<std::__future_base::_Async_state_impl<_BoundFn, <template-parameter-1-2> >::_Async_state_impl(_BoundFn&&) [with _BoundFn = std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>; _Res = void]::__lambda3()> >::_M_run(void) (this=0xf82d58)
    at /usr/include/c++/4.8/thread:115
#23 0x00007ffff6912a40 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#24 0x00007ffff612f182 in start_thread (arg=0x7fffdcfca700)
    at pthread_create.c:312
---Type <return> to continue, or q <return> to quit---
#25 0x00007ffff5e5c47d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

~~~~~

**Comment by jtanx** [comment]
~~~~~
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffdc7c9700 (LWP 3638)]
0x0000000000553cbe in picopter::Observations::lastObservation (this=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/observations.cpp:75
75	    return accumulator.sample_time;
(gdb) p accumulator.sample_time
Cannot access memory at address 0x0
(gdb) bt
#0  0x0000000000553cbe in picopter::Observations::lastObservation (this=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/observations.cpp:75
#1  0x000000000054d6f2 in picopter::ObjectTracker::Run (this=0xfc3e18, 
    fc=0x10322c0, opts=0x0)
    at /home/jeremy/fyp/picopterx/code/src/modules/object_tracker.cpp:270
#2  0x0000000000524c71 in picopter::FlightController::__lambda4::operator() (
    __closure=0xfcebe0)
    at /home/jeremy/fyp/picopterx/code/src/base/flightcontroller.cpp:312
#3  0x0000000000527208 in std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>::_M_invoke<>(std::_Index_tuple<>) (this=0xfcebe0)
    at /usr/include/c++/4.8/functional:1732
#4  0x0000000000527007 in std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()>::operator()(void) (this=0xfcebe0) at /usr/include/c++/4.8/functional:1720
#5  0x0000000000526d46 in std::_Function_handler<void(), std::reference_wrapper<std::_Bind_simple<picopter::FlightController::RunTask(picopter::TaskIdentifier, std::shared_ptr<picopter::FlightTask>, void*)::__lambda4()> > >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/4.8/functional:2101
#6  0x000000000052c68e in std::function<void ()>::operator()() const (
    this=0xfadcf8) at /usr/include/c++/4.8/functional:2471
#7  0x000000000052c4d8 in std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, void>

snip
~~~~~

File: code/src/modules/observations.cpp [hunk 2]
```diff
@@ -11,36 +11,59 @@
...
+
+    accumulator.location = combineDistribs(accumulator.location, observation.location);
+    accumulator.velocity = combineDistribs(accumulator.velocity, observation.velocity);
+    //update the time stamp
+    if(accumulator.sample_time < observation.sample_time) accumulator.sample_time=observation.sample_time;
+    accumulator.source = observation.source;
+}
+void Observations::updateObject(TIME_TYPE timestep){
+
+    accumulator.location = vectorSum(accumulator.location, stretchDistrib(accumulator.velocity, ((double)(timestep.count()))/TICKS_PER_SEC ));
```
Comment (jtanx) [inline code/src/modules/observations.cpp]: Should probably be using `duration_cast<milliseconds>(timestep).count()` or similar

File: code/include/observations.h [hunk 3]
```diff
@@ -75,56 +77,85 @@ namespace picopter {
         CAMERA_FLOW,    //optical flow will resolve a range and blob (probably include two images and IMU velocity)
```
Comment (jtanx) [inline code/include/observations.h]: If you want your comments to show up in documentation, you should use Doxygen style comments, e.g.

~~~~~
/** Optical flow will resolve a range and blob (probably include two images and IMU velocity) **/
~~~~~

File: code/include/observations.h [hunk 4]
```diff
@@ -47,14 +47,16 @@ Motivated by investigation into the monocular case of:
 #include "camera_stream.h"
 #include <opencv2/opencv.hpp>
 
+#define TICKS_PER_SEC 1000000000l
+#define TIME_TYPE std::chrono::duration<long int, std::ratio<1l, TICKS_PER_SEC> >
+#define CLOCK_TYPE std::chrono::time_point<std::chrono::_V2::steady_clock, std::chrono::duration<long int, std::ratio<1l, TICKS_PER_SEC> > >
```
Comment (jtanx) [inline code/include/observations.h]: Can be replaced with

~~~~~
#define CLOCK_TYPE std::chrono::time_point<std::chrono::steady_clock>
~~~~~

File: code/include/observations.h [hunk 5]
```diff
@@ -47,14 +47,16 @@ Motivated by investigation into the monocular case of:
 #include "camera_stream.h"
 #include <opencv2/opencv.hpp>
 
+#define TICKS_PER_SEC 1000000000l
+#define TIME_TYPE std::chrono::duration<long int, std::ratio<1l, TICKS_PER_SEC> >
```
Comment (jtanx) [inline code/include/observations.h]: Can be replaced with 
~~~~~
#define TIME_TYPE std::chrono::steady_clock::duration
~~~~~

If you use `duration_cast` for calling `stretchDistrib`

File: code/src/modules/object_tracker.cpp [hunk 6]
```diff
@@ -423,4 +607,29 @@ void ObjectTracker::CalculateTrackingTrajectory(FlightController *fc, Vec3D *cou
...
+        DEG2RAD(coord.lat - launch_point.lat) * (1000*RADIUS_OF_EARTH),
+        DEG2RAD(coord.lon - launch_point.lon) * (1000*RADIUS_OF_EARTH*cos(RAD2DEG(launch_point.lat))),
+        (launch_point.alt - coord.alt)
+        );
+    return retval;
+}
+Coord3D ObjectTracker::GPSFromGround(Vec3d coord){
+    Coord3D retval;
+    retval.lat = launch_point.lat + RAD2DEG(coord(0)/(1000*RADIUS_OF_EARTH));
+    retval.lon = launch_point.lon + RAD2DEG(coord(1)/(1000*RADIUS_OF_EARTH*cos(RAD2DEG(launch_point.lat))));
```
Comment (jtanx) [inline code/src/modules/object_tracker.cpp]: This function basically duplicates `CoordAddOffset`. 

`cos(RAD2DEG(launch_point.lat))` should be `cos(DEG2RAD(launch_point.lat))`. This also applies to `GroundFromGPS`.

File: code/include/object_tracker.h [hunk 7]
```diff
@@ -32,8 +37,31 @@ namespace picopter {
             void SetTrackMethod(TrackMethod method);
             void Run(FlightController *fc, void *opts) override;
             bool Finished() override;
+
+            cv::Vec3d GroundFromGPS(navigation::Coord3D coord);
+            navigation::Coord3D GPSFromGround(cv::Vec3d coord);
+
+            navigation::Coord3D launch_point;       //centre and orientation of the ground coordinate system
```
Comment (jtanx) [inline code/include/object_tracker.h]: This needs to be set in the object tracker Run function - it's currently not set anywhere which is causing it to have undefined values.

Also ideally this should not be public. If you need to access it from outside, I suggest a getter/setter function.

**Comment by jtanx** [comment]
@BrettRD you can add the following in the `Run` function (around line 153; just after we've checked that we got authorisation):

`````c++
    //Get the launch position.
    if (!fc->fb->GetHomePosition(&launch_point)) {
        Log(LOG_WARNING, "I don't know the launch position! Bailing!!!");
        return;
    }
``````

**Comment by jtanx** [comment]
From my brief testing in the simulator it still looks like the copter will go flying off into the horizon (at 5m/s) with the current gains. The tracking arrow shouldn't be near 100% (extends from the centre to the edge of the frame) when the object is close to centre.

File: code/src/modules/object_tracker.cpp (hunk 6 above)
Comment (BrettRD) [inline code/src/modules/object_tracker.cpp]: The object tracker relies on processes that occur in NED coordinates (requires euclidean space). CoordAddOffset doesn't expose NED vectors

**Pull request merged**